x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/HashLookupOperator.java

```diff
@@ -0,0 +1,138 @@
+import org.elasticsearch.core.ReleasableIterator;
+import org.elasticsearch.core.Releasables;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class HashLookupOperator extends AbstractPageMappingToIteratorOperator {
+    /**
+     * Factory for {@link HashLookupOperator}. It's received {@link Block}s
+     * are never closed, so we need to build them from a non-tracking factory.
+     */
+    public static class Factory implements Operator.OperatorFactory {
+        private final Block[] keys;
+        private final int[] blockMapping;
+
+        public Factory(Block[] keys, int[] blockMapping) {
+            this.keys = keys;
+            this.blockMapping = blockMapping;
+        }
+
+        @Override
+        public Operator get(DriverContext driverContext) {
+            return new HashLookupOperator(driverContext.blockFactory(), keys, blockMapping);
+        }
+
+        @Override
+        public String describe() {
+            StringBuilder b = new StringBuilder();
+            b.append("HashLookup[keys=[");
+            for (int k = 0; k < keys.length; k++) {
+                Block key = keys[k];
+                if (k != 0) {
+                    b.append(", ");
+                }
+                b.append("{type=").append(key.elementType());
+                b.append(", positions=").append(key.getPositionCount());
+                b.append(", size=").append(ByteSizeValue.ofBytes(key.ramBytesUsed())).append("}");
+            }
+            b.append("], mapping=").append(Arrays.toString(blockMapping)).append("]");
+            return b.toString();
+        }
+    }
+
+    private final BlockHash hash;
+    private final int[] blockMapping;
+
+    public HashLookupOperator(BlockFactory blockFactory, Block[] keys, int[] blockMapping) {
+        this.blockMapping = blockMapping;
+        List<BlockHash.GroupSpec> groups = new ArrayList<>(keys.length);
+        for (int k = 0; k < keys.length; k++) {
+            groups.add(new BlockHash.GroupSpec(k, keys[k].elementType()));
+        }
+        /*
+         * Force PackedValuesBlockHash because it assigned ordinals in order
+         * of arrival. We'll figure out how to adapt other block hashes to
+         * do that soon. Soon we must figure out how to map ordinals to rows.
+         * And, probably at the same time, handle multiple rows containing
+         * the same keys.
+         */
+        this.hash = BlockHash.buildPackedValuesBlockHash(
+            groups,
+            blockFactory,
+            (int) BlockFactory.DEFAULT_MAX_BLOCK_PRIMITIVE_ARRAY_SIZE.getBytes()
+        );
+        boolean success = false;
+        try {
+            final int[] lastOrd = new int[] { -1 };
+            hash.add(new Page(keys), new GroupingAggregatorFunction.AddInput() {
+                @Override
+                public void add(int positionOffset, IntBlock groupIds) {
+                    // TODO support multiple rows with the same keys
```

Comment:
This is a *huge* TODO, innocently tucked in here. And it dovetails into the block comment above about using different types for the lookup. I don't know how to solve this one at the moment. So I'm saving it fair a rainy day.